FSW/src/util/encodedFile.cpp

```diff
@@ -0,0 +1,207 @@
+ *  Constructs an empty EncodedFile
+ *  
+ * Inputs:
+ *  None
+ */
+EncodedFile::EncodedFile() {
+    m_timestamp = 0;
+}
+
+/* - - - - - - Constructor (science data) - - - - - - *
+ * Usage:
+ *  Constructs an EncodedFile and fills it with new science data
+ *  
+ * Inputs:
+ *  buffer - pointer to buffer of photodiode data
+ *  timestamp - file timestamp
+ */
+EncodedFile::EncodedFile(uint16_t *buffer, unsigned long &timestamp) : EncodedFile::EncodedFile() {
+    encodeData(buffer, timestamp);
+}
+
+/* - - - - - - Constructor (encoded data) - - - - - - *
+ * Usage:
+ *  Constructs an EncodedFile and fills it with already-encoded data
+ *  
+ * Inputs:
+ *  encodedData - pointer to encoded file data, typecast to void*
+ */
+EncodedFile::EncodedFile(void *encodedData) : EncodedFile::EncodedFile() {
+    fill(encodedData);
+}
+
+/* - - - - - - encodeData - - - - - - *
+ * Usage:
+ *  Fills the file with science data and encodes the file
+ *  
+ * Inputs:
+ *  buffer - pointer to buffer of photodiode data
+ *  timestamp - file timestamp
+ * 
+ * Outputs:
+ *  None
+ */
+void EncodedFile::encodeData(uint16_t *buffer, unsigned long &timestamp) {
+    
+    memset(m_data, 0, ENCODED_FILE_SIZE); // clear the data array
+    memcpy(m_data, buffer, BUFFER_MEMSIZE); // copy buffer to data array
+    memcpy(m_data + BUFFER_MEMSIZE, &timestamp, TIMESTAMP_SIZE); // copy timestamp to data array
+
+    /* encode the data */ 
+    for (int blockNum = 0; blockNum < MESSAGE_COUNT; blockNum++) { // for each block...
+        // copy a chunk of unencoded data into a temporary message block
+        uint8_t message[MESSAGE_SIZE]; 
+        memcpy(message, m_data + blockNum * MESSAGE_SIZE, MESSAGE_SIZE);
+
+        // encode the message
+        m_blocks[blockNum].encodeMessage(message);
+    }
+}
+
+/* - - - - - - fill - - - - - - *
+ * Usage:
+ *  Fills the file with encoded file data
+ *  
+ * Inputs:
+ *  encodedData - pointer to already-encoded file data, typecast to void* 
+ * 
+ * Outputs:
+ *  None
+ */
+void EncodedFile::fill(void *encodedData) {
+    // overwrite each block
+    for (int blockNum = 0; blockNum < MESSAGE_COUNT; blockNum++) { // for each block...
+        m_blocks[blockNum].fill(encodedData + blockNum * HAMMING_BLOCK_SIZE); // overwrite the block with new data
+    }
+}
+
+/* - - - - - - scrub - - - - - - *
+ * Usage:
+ *  Scans the file for errror and corrects single bit errors.
+ *  Corrupted blocks that cannot be corrected are cleared.
+ *  
+ * Inputs:
+ *  None
+ * 
+ * Outputs:
+ *  A ScrubReport struct containing counts of found and corrected errors
+ */
+ScrubReport EncodedFile::scrub() {
+    ScrubReport scrubInfo;
+
+    /* scan and correct each block */
+    for (int blockNum = 0; blockNum < MESSAGE_COUNT; blockNum++) { // for each block...
+        // correct the block and get an error report
+        ErrorReport errorInfo = m_blocks[blockNum].correctBlock();
+
+        // if uncorrectable error detected, clear the block
```

Comment:
What I've seen typically in data science is that faulty data is flagged rather than outright deleted. However, I think when it comes to an error like multiple bit flips - there's no point not deleting it, because there's nothing you could do to fix that data correct? Deleting it also has advantages in terms of computing time and saving memory, so I think this is still the better solution, just curious if you had thought of the alternative.

Reply:
Correct, for a multiple bit error, there is no way to know where the error is, so the whole block is unfixable. I guess we could theoretically implement some super cool system that looks at surrounding data points, interpolates the likely range of values for the block, and determines which bits were likely flipped... that would actually be awesome, but no. When we clear a block like this, it will look like four consecutive measurements of zero irradiance, so right now there is no way to tell the difference between a corrupted block and an uncorrupted block that happens to be all zeros. The issue is we can't flag a block without adding extra bits to the file, since the blocks are already at maximum data density, and since we don't know how many blocks are going to be corrupted in any given file, we would need to append a bunch of redundant bits just to store flags. One alternative I had in mind is to overwrite corrupted blocks with a 64 bit int equal to the index of the block in the file. Then when we decode a file, if any of the blocks contain their own index then we can be fairly confident that block was wiped. The likelihood that a real measurement happens to be that same value is extremely low. Of course, any data we write in the file is also vulnerable to errors, so there is no way to win 100% of the time.

Reply:
Sounds good, glad to see it's on the mind.